Tools/CISupport/ews-build/steps.py

```diff
@@ -2150,6 +2153,189 @@ def hideStepIf(self, results, step):
+            self._addToLog('stdio', 'PRs to block: {}'.format(label_as_blocked))
+
+            if len(label_as_safe):
+                self.build.addStepsAfterCurrentStep([SafeMergeToMergeQueue()])
+            if len(label_as_blocked):
+                self.build.addStepsAfterCurrentStep([SafeMergeToMergeBlocked()])
+        return SUCCESS if num_prs else FAILURE
+
+    def getResultSummary(self):
+        if self.results == SUCCESS:
+            return {'step': f"Successfully requested and filtered pull requests"}
+        elif self.results == FAILURE:
+            return {'step': f"Failed to request and filter pull requests"}
+        return buildstep.BuildStep.getResultSummary(self)
+
+    def queryGraphQL(self, payload):
+        graphql_url = 'https://api.github.com/graphql'
+        username, access_token = GitHub.credentials(user=GitHub.user_for_queue(self.getProperty('buildername', '')))
+        if username and access_token:
+            auth_header = b64encode('{}:{}'.format(username, access_token).encode('utf-8')).decode('utf-8')
+            headers = {'Authorization': 'bearer {}'.format(access_token)}
+
+        # TODO: implement twistd
+        response = requests.post(url=graphql_url, json=payload, headers=headers)
+        data = json.loads(response.text)
+        if 'errors' in response:
+            self._addToLog('stdio', response['errors'])
+        return data
+
+    def getNumPRs(self, label='\"safe-merge-queue\"'):
+        query_body = '{ repository(owner:"Webkit", name:"WebKit") { pullRequests(labels: %s) { totalCount } } }' % label
+        query = {'query': query_body}
+        num_prs = self.queryGraphQL(query)['data']['repository']['pullRequests']['totalCount']
+        if num_prs:
+            yield self._addToLog('stdio', 'There are {} PRs in safe-merge-queue.'.format(num_prs))
+            return num_prs
+        else:
+            # error handing tbd
+            yield self._addToLog('stdio', 'Failed to retrieve number of PRs.')
+            return None
+
+    def sortPRs(self, num_prs, label='\"safe-merge-queue\"'):
+        query_body = '{ repository(owner:"Webkit", name:"WebKit") { pullRequests(labels: %s, last: %s) { edges { node { title number commits(last: 3) { nodes { commit { commitUrl status { state contexts { context state } } } } } } } } } }' % (label, num_prs)
+        query = {'query': query_body}
+
+        self._addToLog('stdio', "Fetching all PRs with label {}.".format(label))
+
+        response = self.queryGraphQL(query)
+        if not response:
+            self._addToLog('stderr', 'Query failed')  # error handling tbd
+            return None
+        all_pr_data = response['data']['repository']['pullRequests']['edges']
+
+        list_of_prs = []
+        passed_status_check = []
+        failed_status_check = []
+        no_checks = []
+
+        ignore_contexts = ['style', 'webkitperl', 'wpe', 'wpe-wk2', 'gtk', 'gtk-wk2', 'api-gtk', 'wincaire']
```

Comment:
I think we want the reverse here: we want to list the tests we require to have passed before landing.

Reply:
Sure, that would just change the proceeding logic a little. But, wouldn't it be preferable if a new test was added and required by default? In the required list implementation, I could see a test added to EWS but not this list and then a PR with an error going into merge-queue.

Reply:
New EWS queues usually start as experimental. The bigger sticking point, though, is that's it's possible for a build to be pending. In that case, we don't want `safe-merge-queue` to be triggered (yet). The pending case is challenging because there may not be a context reported yet.

Reply:
Okay, I'll make the change. I'm currently putting all PRs with no status into its own list that's ignored, perhaps that takes care of pending builds?